benchmark/utils/stencil_matrix.hpp

```diff
@@ -0,0 +1,577 @@
+
+    std::array<int, 2> coords{};
+    coords[0] = comm.rank() % dims[0];
+    coords[1] = comm.rank() / dims[0];
+
+    return generate_2d_stencil_box<ValueType, IndexType>(
+        dims, coords, target_local_size, restricted);
+}
+
+
+/**
+ * Generates matrix data for a given 23 stencil, where the position of this
+ * block is given by it's MPI rank.
+ *
+ * @see generate_3d_stencil_box
+ */
+template <typename ValueType, typename IndexType>
+gko::matrix_data<ValueType, IndexType> generate_3d_stencil(
+    gko::experimental::mpi::communicator comm,
+    const gko::size_type target_local_size, bool restricted)
+{
+    std::array<int, 3> dims{};
+    MPI_Dims_create(comm.size(), dims.size(), dims.data());
+
+    std::array<int, 3> coords{};
+    coords[0] = comm.rank() % dims[0];
+    coords[1] = (comm.rank() / dims[0]) % dims[1];
+    coords[2] = comm.rank() / (dims[0] * dims[1]);
+
+    return generate_3d_stencil_box<ValueType, IndexType>(
+        dims, coords, target_local_size, restricted);
+}
+
+
+/**
+ * Generates matrix data for a 2D stencil matrix. If restricted is set to true,
+ * creates a 5-pt stencil, if it is false creates a 9-pt stencil.
+ *
+ * The degrees of freedom are ordered such that each MPI rank has at most two
+ * neighbors.
+ *
+ * @see generate_2d_stencil_box
+ */
+template <typename ValueType, typename IndexType>
+gko::matrix_data<ValueType, IndexType> generate_2d_stencil_with_optimal_comm(
+    gko::experimental::mpi::communicator comm,
+    const IndexType target_local_size, bool restricted)
+{
+    const auto dp =
+        static_cast<IndexType>(closest_nth_root(target_local_size, 2));
+    const auto mat_size = dp * dp * comm.size();
+    const auto rows_per_rank = gko::ceildiv(mat_size, comm.size());
+    const auto start = rows_per_rank * comm.rank();
+    const auto end = gko::min(rows_per_rank * (comm.rank() + 1), mat_size);
+
+    auto A_data = gko::matrix_data<ValueType, IndexType>(
+        gko::dim<2>{static_cast<gko::size_type>(mat_size),
+                    static_cast<gko::size_type>(mat_size)});
+
+    auto is_valid_neighbor = [&](const IndexType d_i, const IndexType d_j) {
+        return !restricted || ((d_i == 0 || d_j == 0));
+    };
+
+    auto nnz_in_row = [&]() {
+        int num_neighbors = 0;
+        for (IndexType d_i : {-1, 0, 1}) {
+            for (IndexType d_j : {-1, 0, 1}) {
+                if (is_valid_neighbor(d_i, d_j)) {
+                    num_neighbors++;
+                }
+            }
+        }
+        return num_neighbors;
+    };
+    const auto diag_value = static_cast<ValueType>(nnz_in_row() - 1);
+
+    for (IndexType row = start; row < end; row++) {
+        auto i = row / dp;
+        auto j = row % dp;
+        for (IndexType d_i = -1; d_i <= 1; d_i++) {
+            for (IndexType d_j = -1; d_j <= 1; d_j++) {
+                if (is_valid_neighbor(d_i, d_j)) {
+                    auto col = j + d_j + (i + d_i) * dp;
```

Comment:
also need a bounds check here

Reply:
Do you mean besides the one below that?

Reply:
yes, I mean `j + d_j and i + d_i in [0,dim]`

Reply:
you're right, thanks!